### DISCUSSION TITLE
Could ImageMagick support raw file which has padding data?

### DISCUSSION THREAD

**Discussion by uraxurax** [ImageMagick/ImageMagick#6035] [Help] [answered]
### ImageMagick version

 6.9.10-23

### Operating system, version and so on

Ubuntu 20.04.4 LTS

### Description

I have the following raw file which has padding data.

RGB raw image
Width: 1920
Height: 1080
bit : 16bit (Valid bit size 10bit, padding bit : 6bit)
Endian: Little

And I want to get  10 bit ppm file from raw file.

So I have 2 questions.

1) Is there any way to get 10 bit ppm file from raw file?
I tried to use following command and get error.

$convert -depth 10 -size 1920x1080 rgb:10bit_gray_lsb_padding.raw 10bit_gray_lsb_padding.ppm
convert-im6.q16: unexpected end-of-file `10bit_gray_lsb_padding.raw': No such file or directory @ error/rgb.c/ReadRGBImage/242.
convert-im6.q16: no images defined `10bit_gray_lsb_padding.ppm' @ error/convert.c/ConvertImageCommand/3258.

2) Is there any way to get 10 bit ppm file from 16 bit ppm file?
When I tried to following command, I could get 16bit ppm file.

$convert -depth 16 -size 1920x1080 rgb:10bit_gray_lsb_padding.raw 16bit_gray.ppm

However, I couldn't find the way to get 10 bit ppm from 16 bit ppm.
Do you have any advice?

Regards,
hiro

[10bit_gray_lsb_padding.zip](https://github.com/ImageMagick/ImageMagick/files/10665192/10bit_gray_lsb_padding.zip)

**Comment by snibgo** [comment]
Your file 10bit_gray_lsb_padding.raw has 12,441,600 bytes, which is 1920\*1080\*2\*3. So I suppose it has three channels, each of 2 bytes, 16 bits. You say that only 10 of bits are image data, and the other 6 are padding.

```
magick -size 1920x1080 -depth 16 RGB:10bit_gray_lsb_padding.raw -verbose info:
```
This tells us that all pixels have the value 32704 in all channels. So the image is one solid colour. I can't tell from that whether the 6 bits of padding are at the high or low end of each 16-bit value.

If the padding is at the low end, and each padding bit is zero, then no further work is needed. If the padding is not zero, then AND the result with 1111111111000000 to zero the bottom 6 bits. Similarly if the padding is at the high end, AND the top 6 bits and shift left or multiply.

**Reply by uraxurax** [reply]
Dear snibgo,

Thank you for your reply.
And I'm so sorry for the lack of information.
The padding is at the low end and each padding bit is zero.

Could I confirm how to get 10 bit ppm file from raw file?

When I used following command, I cannot get the file by the error.
`
magick -depth 10 -size 1920x1080 rgb:10bit_gray_lsb_padding.raw 10bit_gray_lsb_padding.ppm
`

When I used following command, I can get *not 10 bit ppm* but *16bit ppm*
`
magick -depth 16 -size 1920x1080 rgb:10bit_gray_lsb_padding.raw 16bit_gray.ppm
`

**Comment by snibgo** [comment]
When you use "`-depth 10`", you are telling IM that the file contains 10 bits per pixel value. In this context, `-depth` says nothing about how many of those bits, if any, are padding. IM reads 10 bits for the red channel, then the next 10 bits for the green channel, and so on.

You want it to read 16 bits for each channel, so you need "`-depth 16`".

In each value, the bottom 6 bits of the 16 will be zero. If your IM is Q16, no further action is needed, and your pixel values will be between 0 and binary 11111111 11000000, which is hex FFC0, decimal 65472. If you want the range to be from zero to 65535, then multiply by 65535/65472:
```
magick -depth 16 -size 1920x1080 rgb:10bit_gray_lsb_padding.raw -evaluate Multiply %[fx:65535/65472] 16bit_gray.ppm
```
(If you use Windows BAT, double the % sign.)

Similarly if you want the range from 0 to 1047:
```
magick -depth 16 -size 1920x1080 rgb:10bit_gray_lsb_padding.raw -evaluate Multiply %[fx:1047/65472] 16bit_gray.ppm
```

**Reply by uraxurax** [reply]
Dear Snibgo,

Thank you for your information.
When I tried to use your commands, I got errors as follows.

`
$convert -depth 16 -size 1920x1080 rgb:10bit_gray_lsb_padding.raw -evaluate Multiply %[fx:65535/65472] 16bit_gray.ppm
`
convert-im6.q16: invalid argument for option `-evaluate': %[fx:65535/65472] @ error/convert.c/ConvertImageCommand/1493.

`
convert -depth 16 -size 1920x1080 rgb:10bit_gray_lsb_padding.raw -evaluate Multiply %[fx:1047/65472] 16bit_gray.ppm
`
convert-im6.q16: invalid argument for option `-evaluate': %[fx:1047/65472] @ error/convert.c/ConvertImageCommand/1493.

Could you give me any advise?

**Reply by fmw42** [reply]
On Linux, you need to put double quotes around %[fx:...]

**Reply by fmw42** [reply]
Also %[fx:...] may not be valid in IM 6 as a value for -evaluate multiply. If that is the case, just do the division manually and put that value there.

**Reply by uraxurax** [reply]
Dear fmw42,

Thank you for your information.
I understood that my environment cannot use %[fx:] option.
I will do the division manually and put that value there.

Regards,
hiro

**Comment by snibgo** [comment]
As fmw42 says: v6 `convert` can't have fx in `-evaluate`, so use "`-evaluate Multiply 1.00096224340176`" etc.

**Reply by uraxurax** [reply]
Dear snibgo,

Thank you for your information.

I can get the exact  value (0x1FF) by using **-evaluate Multiply 0.015625** option as follows.

`
convert -depth 16 -size 1920x1080 rgb:10bit_gray_lsb_padding.raw -evaluate Multiply 0.015625 10bit_gray.ppm
`

However, the output file seems to be still 16bit ppm.

$ identify 10bit_gray.ppm 
10bit_gray.ppm PPM 1920x1080 1920x1080+0+0 16-bit sRGB 11.8653MiB 0.030u 0:00.019

Is there any solution?

[10bit_gray.zip](https://github.com/ImageMagick/ImageMagick/files/10673779/10bit_gray.zip)

**Comment by snibgo** [comment, marked as answer]
The binary PPM format records values in either one byte or in two bytes. It has no mechanism for packing values in other than 8 or 16 bits. However, you can "`-depth 10`" before writing the output. This will write metadata that says the maximum value is 1023.

**Reply by uraxurax** [reply]
Dear snibgo,

Thank you for your kind support.
I can get 10 bit ppm by your advice.

`
$convert -depth 16 -size 1920x1080 rgb:10bit_gray_lsb_padding.raw -evaluate Multiply 0.015625 -depth 10 10bit_gray.ppm
`

```
$ identify 10bit_gray.ppm 
10bit_gray.ppm PPM 1920x1080 1920x1080+0+0 10-bit sRGB 11.8653MiB 0.030u 0:00.039
```

Regards,
hiro
